File: 2021f/cs203/lab3/F.java
import java.io.BufferedReader;
import java.util.StringTokenizer;
import java.io.*;
import java.util.*;

public class F {
    static Plant[] tarr;

    public static void main(String[] args) {
        QReader sc = new QReader();
        int n = sc.nextInt();
        int p = sc.nextInt();
        int q = sc.nextInt();
        Plant[] ps = new Plant[n];
        tarr = new Plant[n];
        long ans = 0;
        int dfc = 0;
        for (int i = 0; i < n; i++) {
            ps[i] = new Plant(sc.nextLong(), sc.nextLong());
            ans += ps[i].st;
            if (ps[i].df > 0)
                dfc++;
        }

        if (q == 0) {
            System.out.println(ans);
            return;
        }
        ms(ps, 0, n - 1);
        if (p == 0) {
            for (int t = 0; t < Math.min(dfc, q); t++)
                ans += ps[t].df;
            System.out.println(ans);
            return;
        }

        if (q > dfc) {
            long maxAdv = 0;
            for (int i = 0; i < n; i++)
                maxAdv = Math.max(maxAdv, (long) Math.pow(2, p) * ps[i].ht - ps[i].st - Math.max(0, ps[i].df));
            for (int i = 0; i < dfc; i++)
                ans += ps[i].df;
            ans += maxAdv;
            System.out.println(ans);
            return;
        }

        long maxAdv = 0;
        for (int i = 0; i < q; i++)
            maxAdv = Math.max(maxAdv, (long) Math.pow(2, p) * ps[i].ht - ps[i].st - ps[i].df);
        for (int i = q; i < n; i++)
            maxAdv = Math.max(maxAdv, (long) Math.pow(2, p) * ps[i].ht - ps[i].st - ps[q - 1].df);
        for (int i = 0; i < q; i++)
            ans += ps[i].df;
        ans += maxAdv;
        System.out.println(ans);
    }

    static void mer(Plant[] arr, Plant[] tmp, int lo, int mi, int hi) {
        int i = lo, j = mi + 1, f = 0;
        while (i <= mi && j <= hi) {
            if (arr[i].df > arr[j].df)
                tmp[f++] = arr[i++];
            else
                tmp[f++] = arr[j++];
        }
        while (i <= mi)
            tmp[f++] = arr[i++];
        while (j <= hi)
            tmp[f++] = arr[j++];
        for (i = lo; i <= hi; i++)
            arr[i] = tmp[i - lo];
    }

    static void ms(Plant[] arr, int lo, int hi) {
        if (lo >= hi)
            return;
        int mi = (lo + hi) / 2;
        ms(arr, lo, mi);
        ms(arr, mi + 1, hi);
        mer(arr, tarr, lo, mi, hi);
    }
}

class Plant {
    long ht, st, df;

    Plant(long ht, long st) {
        this.ht = ht;
        this.st = st;
        this.df = ht - st;
    }
}

class QReader {
    private BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    private StringTokenizer tokenizer = new StringTokenizer("");

    private String innerNextLine() {
        try {
            return reader.readLine();
        } catch (IOException e) {
            return null;
        }
    }

    public boolean hasNext() {
        while (!tokenizer.hasMoreTokens()) {
            String nextLine = innerNextLine();
            if (nextLine == null) {
                return false;
            }
            tokenizer = new StringTokenizer(nextLine);
        }
        return true;
    }

    public String nextLine() {
        tokenizer = new StringTokenizer("");
        return innerNextLine();
    }

    public String next() {
        hasNext();
        return tokenizer.nextToken();
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

    public long nextLong() {
        return Long.parseLong(next());
    }
}

File: 2021f/cs203/lab3/f.cpp
#include <math.h>

#include <iostream>

using namespace std;

struct plant {
    long hi;
    long st;
    long df;  // hi-st
};

plant *tmp = nullptr;

void mer(plant *arr, int lo, int mi, int hi) {
    int i = lo, j = mi + 1, f = 0;
    while (i <= mi && j <= hi) {
        if (arr[i].df < arr[j].df)
            tmp[f++] = arr[j++];
        else
            tmp[f++] = arr[i++];
    }
    while (i <= mi) tmp[f++] = arr[i++];
    while (j <= hi) tmp[f++] = arr[j++];
    for (int n = lo; n <= hi; n++) arr[n] = tmp[n - lo];
}

void msp(plant *arr, int lo, int hi) {
    if (lo >= hi) return;
    int mi = (lo + hi) / 2;
    msp(arr, lo, mi);
    msp(arr, mi + 1, hi);
    mer(arr, lo, mi, hi);
}

int main() {
    int n, p, q, shuBc = 0;
    scanf("%d %d %d", &n, &p, &q);
    plant *ps = new plant[n];
    tmp = new plant[n];
    for (int i = 0; i < n; i++) {
        scanf("%ld %ld", &ps[i].hi, &ps[i].st);
        ps[i].df = ps[i].hi - ps[i].st;
        if (ps[i].df > 0) shuBc++;
    }
    if (!q) {
        unsigned long long ans = 0;
        for (int i = 0; i < n; i++) ans += ps[i].st;
        printf("%llu\n", ans);
        return 0;
    }
    unsigned long long mx = 0, ans = 0;
    msp(ps, 0, n - 1);
    if (!p) {
        if (q > shuBc) {
            for (int i = 0; i < n; i++) ans += ps[i].st + max(0l, ps[i].df);
        } else {  // big df, &max del df.ndf
            for (int i = 0; i < n; i++) ans += ps[i].st;
            for (int i = 0; i < q; i++) ans += ps[i].df;
        }
        printf("%llu\n", ans);
        return 0;
    }
    if (q > shuBc) {
        for (int i = 0; i < n; i++) {
            mx = max(mx,  // origin
                     ps[i].hi * (unsigned long long)pow(2, p) -
                         (ps[i].st + max(0l, ps[i].df)));  // delta: ndf, df
            ans += ps[i].st + max(0l, ps[i].df);
        }
        ans += mx;
        printf("%llu\n", ans);
        return 0;
    } else {  // big df, &max del df.ndf
        for (int i = 0; i < q; i++)
            mx = max(mx, (unsigned long long)ps[i].hi *
                                 (unsigned long long)pow(2, p) -
                             ps[i].st - ps[i].df);
        // diff contribution
        for (int i = q; i < n; i++)
            mx = max(mx, (unsigned long long)ps[i].hi *
                                 (unsigned long long)pow(2, p) -
                             ps[i].st - ps[q - 1].df);
        // if add this, ndf=nhi-st-oldf // must 0..q-2 df & 1 ndf
        for (int i = 0; i < n; i++) ans += ps[i].st;
        for (int i = 0; i < q; i++) ans += ps[i].df;
        ans += mx;
        printf("%llu\n", ans);
        return 0;
    }
}
